2020/11/seating/src/main.rs

```diff
@@ -0,0 +1,411 @@
+                break;
+            } else if seat == Position::Empty {
+                break;
+            }
+        }
+
+        let mut m = min(x, y);
+        for d in 1..m+1 {
+            let seat = self.map[y-d][x-d];
+            // println!("  Looking up/left at ({},{}), found {:?}", y-d, x-d, seat);
+            if seat == Position::Occupied {
+                // print!("up/left ");
+                occupied += 1;
+                break;
+            } else if seat == Position::Empty {
+                break;
+            }
+        }
+
+        m = min(x + 1, self.map.len() - y);
+        for d in 1..m {
+            let seat = self.map[y+d][x-d];
+            if seat == Position::Occupied {
+                // print!("down/left ");
+                occupied += 1;
+                break;
+            } else if seat == Position::Empty {
+                break;
+            }
+        }
+
+        m = min(self.map[y].len() - x, y + 1);
+        for d in 1..m {
+            let seat = self.map[y-d][x+d];
+            // println!("  Checking ({}, {}), seat is {:?}", y-d, x+d, seat);
+            if seat == Position::Occupied {
+                // print!("up/right ");
+                occupied += 1;
+                break;
+            } else if seat == Position::Empty {
+                break;
+            }
+        }
+
+        m = min(self.map[y].len() - x, self.map.len() - y);
+        for d in 1..m {
+            let seat = self.map[y+d][x+d];
+            if seat == Position::Occupied {
+                // print!("down/right ");
+                occupied += 1;
+                break;
+            } else if seat == Position::Empty {
+                break;
+            }
+        }
+
+        // println!("Found {} occupied", occupied);
+        occupied
+    }
+
+    fn one_if_occupied(&self, y: isize, x: isize) -> usize {
+        match self.get_with_bounds(y, x) {
+            Position::Occupied => 1,
+            _ => 0,
+        }
+    }
+
+    fn get_with_bounds(&self, y: isize, x: isize) -> Position {
+        if y < 0 || y as usize >= self.map.len() {
+            return Position::Floor;
+        }
+
+        if x < 0 || x as usize >= self.map[y as usize].len() {
+            return Position::Floor;
+        }
+
+        return self.map[y as usize][x as usize];
+    }
+
+    pub fn find_steady_state(&self, adjacent: bool) -> Layout {
+        let mut prev = self.clone();
+        loop {
+            println!("Current layout is:");
+            println!("{:?}", prev);
+            let cur = prev.apply_round(adjacent);
+
+            if prev == cur {
+                return cur;
+            }
+
+            prev = cur;
+        }
+    }
+
+    pub fn occupied(&self) -> usize {
+        let mut count = 0;
+        for x in &self.map {
+            for y in x {
+                count += match y {
+                    Position::Occupied => 1,
+                    _ => 0,
+                }
+            }
+        }
```

Comment:
Definitely some combinators on `Iterator` that could make this functiony :). The lack of generics and thus any useful functional-programming operators is driving me nuts doing this in Go :(.